### PR TITLE
Assignment 4

### PR DESCRIPTION
**Pull request by Fervidusletum**
Completed

### CODE REVIEW

**Comment by delta-plus** [comment]
I will review this.

**Review by delta-plus — APPROVED** [review]
Solid work! Looks good to me.

    Create a UserController in the API projects
        Implement endpoints for:
            Retrieving all users ✔
            Retriving a single user by id ✔
            Creating a user ✔
            Updating a user ✔
            Deleting a user ✔
        The UserController should function. ✔
        Include a .rest file for interacting with the controller (REST Client). ✔
    The IUserRepository interface is provided for you. You will need to provide an implementation of this interface within the SecretSanta.Business project. This implementation can simply store the user within a list, dictionary, array, or similar. ✔
    All test projects should be within a test directory ✔
    Create a unit test project for the SecretSanta.Business project. ✔
        Unit test all public methods ✔
        Ensure the project is named with a ".Tests" suffix ✔
        Ensure test class is named appropriately for the class being tested: ✔
        Name all tests in the form <MethodBeingTested>_<ConditionBeingTested>_<ExpectedOutcome>(). You can add <NoWarn>CA1707;INTL0003</NoWarn> to remove the naming warnings. ✔
    Create a Unit Test project for the SecretSanta.Api project. ✔
        Test all controller public methods. It is up to you to determine the best way to provide the controller's dependencies. ✔
        Ensure the project is named with a ".Tests" suffix ✔

File: SecretSanta/src/SecretSanta.Api/Dto/DtoUser.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿namespace SecretSanta.Api.Dto
+{
+    public class DtoUser
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Api/Dto/DtoUser.cs:3]: Nicely done!

File: SecretSanta/src/SecretSanta.Api/Startup.cs [hunk 2]
```diff
@@ -16,6 +14,9 @@ public class Startup
         // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
         public void ConfigureServices(IServiceCollection services)
         {
+            services.AddSingleton<ICollection<User>, TestCollection>();
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Api/Startup.cs:17]: Clever use of DI.

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/UsersControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            mock.GetItemUser = null!;
+            UsersController sut = new(mock);
+
+            ActionResult result = sut.Put(0, new DtoUser());
+
+            Assert.IsTrue(result is NotFoundResult);
+        }
+
+
+        private class mockRepo : IUserRepository
```
Comment (Keboo) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/UsersControllerTests.cs:148]: Nit: Classes should be named with PascalCase even when they are private.
```suggestion
        private class MockRepo : IUserRepository

```

**Review by Keboo — COMMENTED** [review]
Really clean code. Good work.

- Create a UserController in the API projects
  - Implement endpoints for:
    - Retrieving all users ✔
    - Retriving a single user by id ✔
    - Creating a user ✔
    - Updating a user ✔
    - Deleting a user ✔
  - The UserController should function. ✔
  - Include a `.rest` file for interacting with the controller. ✔
- The `IUserRepository` interface is provided for you. You will need to provide an implementation of this interface within the SecretSanta.Business project. This implementation can simply store the user within a list, dictionary, array, or similar. ✔
- All test projects should be within a test directory ✔
- Create a unit test project for the SecretSanta.Business project. ✔
  - Unit test all public methods ✔
  - Ensure the project is named with a ".Tests" suffix ✔
  - Ensure test class is named appropriately for the class being tested: ✔
  - Name all tests in the form `<MethodBeingTested>_<ConditionBeingTested>_<ExpectedOutcome>()`. You can add  `<NoWarn>CA1707;INTL0003</NoWarn>` to remove the naming warnings. ✔
- Create a Unit Test project for the SecretSanta.Api project. ✔
  - Test all controller public methods. It is up to you to determine the best way to provide the controller's dependencies. ✔
  - Ensure the project is named with a ".Tests" suffix ✔

**Pull request closed without merging**